Day 2- updates based on dry run (#50)
AlexsLemonade/training-modules: scRNA-seq/01-normalizing_scRNA-seq.Rmd

```

Let's look at how the gene means look in this dataset. 
We will use `apply` in order to calculate things across our data.frame. 
The second argument in `apply` specifies whether we are calculating by rows or columns. 
(1 = rows, 2 = columns).

```{r}
# Let's calculate the gene means and make a density plot
gene_means <- apply(sc_data, 1, mean)
# What do you think the dimensions of `gene_means` are? 
```

Let's make our first density plot with these data.

```{r}
# Plot the density of the means using ggplot2's `qplot`
qplot(gene_means, geom = "density", xlab = "Gene Mean Counts") 
```

The counts data has so many zeroes, which we should expect. 
83.3% of the data points are zeroes. 

## Filter the counts matrix
  
The small amount of RNA in a single cell results in higher chances of errors or 
biases in amplification steps, so we don't necessarily want to keep the data for 
all the cells and genes

#### Total counts as a quality measure

Let's take a look at this data from how many counts are there per sample. 
We will use `apply` again here, but what will we use for the second argument?

```{r}
# Make a vector of the total number of counts per sample

> total_counts <- apply(sc_data, 2, sum)

Reviewer: Do you want to do
```suggestion
total_counts <- apply(sc_data, , sum)
```
instead? Or perhaps

```suggestion
total_counts <- apply(sc_data, <ROW OR COLUMN INDEX>, sum)
```

Author: What about this? I tried to tell R Markdown to `eval=FALSE` the chunk, but it still had problems with the <ROW OR COLUMN INDEX> portion. 

We will use `apply` again here, but what will we use for the second argument?
Copy and paste this code then replace `<ROW_OR_COLUMN_INDICATOR>` with the correct
argument. 
`total_counts <- apply(sc_data, <ROW_OR_COLUMN_INDICATOR>, sum)`

```{r}
# Make a vector of the total number of counts per sample using the above
# apply template
```

Author: Figured out an alternative. Because `total_counts` is used for a bunch of downstream steps, it makes it too complicated to do this here, instead, I changed it to have this 'fill in the blank' be for the `gene_means` step.

```{r}
# Make a vector that contains the number of genes expressed by a particular cell
num_genes_exp <- apply(detection_mat, 2, sum)
```

Let's plot this using the same style and type of graph as above. 

```{r}
# Let's take a look at what this looks like:
qplot(num_genes_exp, geom = "density", xlab = "Number of Genes Expressed by Each Sample") +
  geom_density(fill = "lightblue") + 
  theme_classic() # Adding this to make it a prettier style aesthetically
```

But as you may have noticed in our `sc_metadata` we have cell-type information 
for these data. 
Let's use the power of our single-cell resolution data to our advantage and look 
at this data with cell-type labels. 
In order to do this we need to prepare a dataframe that has our `num_genes_exp`
numbers along side our cell-type labels.

```{r}
# Let's make a dataframe with this information
num_genes_exp_df <- data.frame(num_genes_exp, 
                               'cell_type' = sc_metadata_filtered$cell.type.ch1) 
```

The next graph we are using the same steps as before but to make individual graphs for each 

> cell type, we are using a fancy option in ggplot2 called `facet_wrap`.

Reviewer: ```suggestion
cell type, we are using an option in ggplot2 called `facet_wrap`.
```

tricky.
So in order to account for these zeroes, we normalize cells in groups with other 
cells like them; a method introduced in [Lun et al 2016](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-016-0947-7).

*These steps may take a minute or so.*

```{r Normalize Data}
# Set seed for reproducibility
set.seed(1234)

> # Step 1) Group cells with other like cells by clustering.  

Reviewer: This seems better :+1:
